### ISSUE TITLE
Parsing lots of data to Realm, causes crashes and null objects.

### ISSUE THREAD

**Issue by MiralDesai** [realm/realm-java#3112]
#### Goal

> What do you want to achieve?
I have a big json file when a user first logs into my app, grabbing all their data, and I want to parsed into multiple RealmObjects. I'm having issues where the data is not being parsed inside an `AsyncTask`, and if I remove the `AsyncTask` the app crashes. Possibly because parsing has not finished and I've already moved onto loading the main activity in my app to use the object? This is all quite hard to explain because I'm not sure if this is an issue with `Realm` or if I need help in figuring out a good structure for parsing that realm likes.

#### Expected Results

> User logs in. Api call, data is retrieved and parsed. At the same time the user is going through the process of loading the main activity where I create a navigation drawer. Everything should load fine.

#### Actual Results

>  Two outcomes depending on how my code is set up.

**Outcome 1:** 
My main entry point for parsing the user data:

```
@Override
    public int saveObjectFromJsonString(final Context context, String jsonString) throws JSONException, SpleatApiException {
        int id = super.saveObjectFromJsonString(jsonString); //This is will save the user data
        final JSONObject json = new JSONObject(jsonString);
        if (json.has(JSON_KEY_TOKEN)) {
            new AsyncTask<Void, Void, Void>() {
                @Override
                protected Void doInBackground(Void... params) {
                    try {
                        saveUserExtrasFromJson(context, json); //This goes off to other classes to parse other user data, such as a users payments for example, so I have a payment object for that.
                    } catch (JSONException e) {
                        e.printStackTrace();
                    }
                    return null;
                }
            };
        }
        return id;
    }
```

When I do the above, the user data like their name, parses fine, but all the extra arrays in json do not parse at all. So my app does not crash, but there is missing data. Objects are null, the payments `RealmObject` I mentioned above for example is null. It's as if the parsing is getting skipped if it's inside an `AsyncTask`.

**Outcome 2:**
```
@Override
    public int saveObjectFromJsonString(final Context context, String jsonString) throws JSONException, SpleatApiException {
        int id = super.saveObjectFromJsonString(jsonString); //parse main user object
        final JSONObject json = new JSONObject(jsonString);
        if (json.has(JSON_KEY_TOKEN)) {
            saveUserExtrasFromJson(context, json); //parse extra arrays within json, same as above.
        }
        return id;
    }
```

So I remove the `AsyncTask` because it looks like the data was being skipped anyway. However this causes my app to crash. Now when I load my main activity (where I have a navigation drawer created) the user object is totally null. I can't get the user's first name or anything.

#### Other comments
I just want to understand why this is happening. By using the `AsyncTask`, it's as if the data is ignored, and when I put breakpoints there, the breakpoints are in fact not hit. So I find it very strange. Is there something with `Realm` I should follow when parsing? Does `Realm` not work with AsyncTasks? 

One other weird thing is that I have a smaller API call that I will be moving over to at a later date(I want to get `Realm` working with my app first) and I gave it a try, and it worked. It doesn't contain some of the bigger arrays, like payments, but it still has some, so I'm wondering if `Realm` just can't handle lots of data at once?

I would love to be able to solve this but I feel a bit stuck at the moment. My code is identical to when I was parsing to an `SQLite` database (without AsyncTask) and there were no crashes or missing data. Anyway would just love to hear the input of others who have spent more time on `Realm`. Thanks.

#### Version of Realm and tooling
Realm version(s): 1.0.0

Android Studio version: 2.1.2

Which Android version and device: Nexus 5, Nexus 6P, Marshmallow, Android N DP4. Not device specific.

FYI this might be slightly related to my previous issue: https://github.com/realm/realm-java/issues/3047 however I tried the suggestion there and I'm still getting issues. Back then I didn't really solve my issue, I just moved to using the small api which I should not have done.

**Comment by Zhuinden** [comment]
Realm can handle 4000 objects or more just fine at a time, if you run the transaction on a background thread and listen to changes of the table with realm change listener.

**Comment by stk1m1** [comment]
@MiralDesai 

Could you please share your crash message with us? It would help us track down the exact cause to support you. Thanks!

**Comment by MiralDesai** [comment]
Hi,

I have since updated to 1.1.0 and I'm using insertOrUpdate() methods now.

The crash I'm getting is a simple NPE when trying to get call the User object and get the firstName variable which is String. the User object itself is null. This is the crash I get when I do not use an AsyncTask, possibly because it's parsing all the data first before making it available to me? If I use an AsyncTask I do not get a crash, instead I just have a User object, but my other objects are empty, as if the data just did not parse.

This is all just a bit confusing to honest.

**Comment by Zhuinden** [comment]
You need to initialize the `User` object first, like `User user = new User()`.

Are you by chance using `User user = realm.createObject(User.class)`?

**Comment by MiralDesai** [comment]
Nope sorry. 

I'm doing it like this:

```
User user = new User();
user.firstname = json.get(firstname); //and for all the others.
RealmUtils.copyOrUpdateObject(realm, user);
```

That utility method does the following:

```
try {
            realm.beginTransaction();
            realm.insertOrUpdate(user);
            //realm.copyToRealmOrUpdate(user);
            realm.commitTransaction();
        } finally {
            realm.close();
        }
```

**Comment by Zhuinden** [comment]
Just make sure you don't use that `RealmUtils` method twice in a row, considering it immediately closes the Realm instance.

Anyways, if that's not the issue, then you'll have to send code over to the Realm people.

**Comment by MiralDesai** [comment]
Right before I do `User user = new User()` I also do `Realm realm = Realm.getDefaultInstance();` so wouldn't that negate that problem?

I definitely make the call multiple times in a short space of time, so the data code does get called multiple times right after each other.

Thanks for your help, I'll see what I can do in the meantime and if need be send my code over to Realm as you suggest.

**Comment by MiralDesai** [comment]
Ok I've actually fixed my issue. Lots of debugging and I realised that my data was being overwritten because the user ids coming down in the json are the same but in fact, the json itself is not complete on one call, and it was parsing in the same place. Instead I doing User user = new User() everytime, I am now checking if a user exists first, then ensuring I do not overwrite data with null if it keys don't exist. 

Very strange circumstances because even I didn't know that the json was different.
